### PR TITLE
feat: Implementing snoostorm polling

### PR DESCRIPTION
**Pull request by callumc34**
# Changes
- Added snoostorm polling for the bot to automatically run when a post is reported
- Added support for the discord bot (see callumc34/discord-bot branch)
- Added support for tagging posts with bad sources (see callumc34/discord-bot branch)
- Reworked the codebase into classes

### CODE REVIEW

File: discord/commands/delete/index.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+            "delete",
+            ["id"],
+            ["ADMINISTRATOR"]
+            );
+    }
+
+    async _run(ctx, args) {
+        return ctx.channel.send("Deleting post...");
+    }
+}
```
Comment (eggdev) [inline discord/commands/delete/index.js:15]: This is something I'd like to see become a single function

File: discord/index.js [hunk 2]
```diff
@@ -1,21 +1,15 @@
...
-// };
+const { NodeJSBot } = require("nodejsdiscordbot");
+
+const KnicksDiscordBot = class KnicksDiscordBot extends NodeJSBot {
+    constructor(prefix, options = {}) {
+        super(prefix, options);
+        this.once("ready", () => {
+            this.commandCollection.loadCommands(process.cwd() + "/discord/commands");
+            console.log("Discord Bot ready...");
+        })
```
Comment (eggdev) [inline discord/index.js]: What exactly is happening here? On creation of KnicksDiscordBot you fire the "ready" command? 

From here what does this class actually do? 

If there are no other bound methods that can be used, it seems like something that could also become a function

File: index.js [hunk 3]
```diff
@@ -2,10 +2,19 @@ const express = require('express');
...
 const port = process.env.PORT || 5000;
 
+//Bot setup
+const clientId = process.env.CLIENT_ID;
+const clientSecret = process.env.CLIENT_SECRET;
+const refreshToken = process.env.REFRESH_TOKEN;
+const subreddit = process.env.SUBREDDIT;
+const badSourceFlairID = process.env.BAD_SOURCE_FLAIR_ID;
+
+bot = new KnicksRedditBot(clientId, clientSecret, refreshToken, subreddit)
```
Comment (eggdev) [inline index.js]: 1) I'd never ever define a variable without let/const. The default should be const in JS as well.

2) Is this being used again anywhere? Or is the plan to use it again in the future?

File: reddit/index.js [hunk 4]
```diff
@@ -1,47 +1,113 @@
...
+    //Snoostorm streams setup
+    this.modQueueStream = new ModQueueStream(this, {
+      subreddit,
+      limit: 50,
+      pollTime: 2000
+    });
+    this.modQueueStream.on("item", async (submission) => {
+      await this.checkSubmission(submission);
+    });
+
```
Comment (eggdev) [inline reddit/index.js]: I'd move a lot of these out of the constructor. Despite the fact that they are needed on instantiation, it should be yanked out into an "init" method that is in the class. The constructor should just "set up" the class, but something else should make it usable

**Review by eggdev — CHANGES_REQUESTED** [review]
All around pretty good stuff with some small changes that I think could be done. Mainly clean up the couple one-off classes you have and make them functional, and let the constructor be isolated from the actual innerworkings of the class

File: index.js (hunk 3 above)
Comment (callumc34) [reply index.js]: 1. Forgot to use const -- fixed
2. Essentially it is the entire program as when its instantiated it starts polling and runs all the checking.
Wasn't really too sure how to implement it.

File: discord/index.js (hunk 2 above)
Comment (callumc34) [reply discord/index.js]: "ready" is discord.js confirming that the bot is logged in and working and so then it loads all the commands.
ready isn't fired until the bot is logged in.

The class currently does very little but the aim is to add functions to it when the discord bot needs to do something.

File: reddit/index.js (hunk 4 above)
Comment (callumc34) [reply reddit/index.js]: Moved stream setup and discord bot setup to their own functions

File: discord/commands/delete/index.js (hunk 1 above)
Comment (callumc34) [reply discord/commands/delete/index.js:15]: The way I coded the NodeJSDiscord bot package means that each command has to be its own class.
I could rework the library to just accept functions if necessary.

Comment (eggdev) [reply discord/commands/delete/index.js:15]: I think libraries should be built so the installed tool can use it however they want. You should never dictate how someone else writes their code (ironic coming from this comment). But you should make the consuming tool have the lowest bar of entry possible

File: index.js (hunk 3 above)
Comment (eggdev) [reply index.js]: If it's going to be a reused variable, I'd keep it in the running app.listen. If for some reason app crashed, and the bot was still running, that'd be bad and you'd have no way to kill it (I guess I'm throwing an edge case out there)

Comment (callumc34) [reply index.js]: Moved it to app.listen

File: discord/commands/delete/index.js (hunk 1 above)
Comment (callumc34) [reply discord/commands/delete/index.js:15]: I built it that way to allow the bot to access more than just the ctx and args passed from discord.js. With functions there would be no signal passed for the main class, KnicksRedditBot, to interact with and then delete the actual post.

I don't think its dictating how you write the code as you could easily add a different method to the class to interact with discord messages. It's more of a guideline.

Comment (eggdev) [reply discord/commands/delete/index.js:15]: What I meant was more that I think that 3rd party imported libraries should be able to be consumed by our application however we want, through Functions or Classes and not be reliant on the libraries way to do things. This is less of a request to change this code, but more of one to make the package a bit more flexible.

> each command has to be its own class.

Like the consuming application may have no interest in using a class and they would immediately not use your package, you know what I mean? Less criticism about what is written here and more best practice around making an open source tool (potentially)

File: .env.example [hunk 5]
```diff
@@ -6,4 +6,7 @@ EXPIRES_IN="from reddit-oauth-helper"
 REFRESH_TOKEN="from reddit-oauth-helper"
 SCOPE="from reddit-oauth-helper"
 SUBREDDIT=knicklejerk
-WEBHOOK_URL="In discord > Server Settings > Integrations > Webhooks > ReportBot"
\ No newline at end of file
+BAD_SOURCE_FLAIR_ID="Flair ID for your bad sources"
```
Comment (eggdev) [inline .env.example:9]: I just tried to spin up the app and caught this. I think we can have these values live in a constants file and not as an environment variable. Probably going to be a lot better in the long run if we are committing them as I'm sure we'll have more values that we'll want to do this with

**Review by eggdev — APPROVED** [review]
Nice work man! I got it spun up and to send a message on a new report to the server so everything seems to be up and running properly!

The couple things I'd like to change up over the course of this are pulling out the Source Flair IDs to being in a constant and possibly even channel ids/bot prefix coming from a config. We may want at some point to send a different bot name to a different channel in regards to something that happens on the sub. 

Also having that committed will help in the long run of adding people to the app.

**Pull request merged**